src/gpuarray_array_blas.c

```diff
@@ -625,50 +622,90 @@ int GpuArray_rgemmBatch_3d(cb_transpose transA, cb_transpose transB, double alph
+                          Bp->data, Bp->offset/elsize, ldb, Bp->strides[0]/elsize,
+                          (float)beta,
+                          Cp->data, Cp->offset/elsize, ldc, Cp->strides[0]/elsize,
+                          batchCount, 0);
     break;
   case GA_FLOAT:
-    err = gpublas_sgemmBatch(o, transA, transB, m, n, k, (float)alpha,
-                             A_datas, A_offsets, lda,
-                             B_datas, B_offsets, ldb,
-                             (float)beta,
-                             C_datas, C_offsets, ldc, batchCount, 0);
+    err = gpublas_sgemm3D(o, transA, transB, m, n, k, (float)alpha,
+                          Ap->data, Ap->offset/elsize, lda, Ap->strides[0]/elsize,
+                          Bp->data, Bp->offset/elsize, ldb, Bp->strides[0]/elsize,
+                          (float)beta,
+                          Cp->data, Cp->offset/elsize, ldc, Cp->strides[0]/elsize,
+                          batchCount, 0);
     break;
   case GA_DOUBLE:
-    err = gpublas_dgemmBatch(o, transA, transB, m, n, k, (double)alpha,
-                             A_datas, A_offsets, lda,
-                             B_datas, B_offsets, ldb,
-                             (double)beta,
-                             C_datas, C_offsets, ldc, batchCount, 0);
+    err = gpublas_dgemm3D(o, transA, transB, m, n, k, (double)alpha,
+                          Ap->data, Ap->offset/elsize, lda, Ap->strides[0]/elsize,
+                          Bp->data, Bp->offset/elsize, ldb, Bp->strides[0]/elsize,
+                          (double)beta,
+                          Cp->data, Cp->offset/elsize, ldc, Cp->strides[0]/elsize,
+                          batchCount, 0);
     break;
   }
 
+  if (err == GA_DEVSUP_ERROR) {
+    gpudata **A_datas = NULL, **B_datas = NULL, **C_datas = NULL;
+    size_t *A_offsets = NULL, *B_offsets = NULL, *C_offsets = NULL;
+    size_t i;
+
+    A_datas = (gpudata**)malloc(batchCount * sizeof(gpudata*));
+    B_datas = (gpudata**)malloc(batchCount * sizeof(gpudata*));
+    C_datas = (gpudata**)malloc(batchCount * sizeof(gpudata*));
+
+    A_offsets = (size_t*)malloc(batchCount * sizeof(size_t));
+    B_offsets = (size_t*)malloc(batchCount * sizeof(size_t));
+    C_offsets = (size_t*)malloc(batchCount * sizeof(size_t));
+
+    if (A_datas == NULL || B_datas == NULL || C_datas == NULL ||
+        A_offsets == NULL || B_offsets == NULL || C_offsets) {
+      err = error_sys(ctx->err, "malloc");
+      goto old_cleanup;
+    }
+
+    for (i = 0; i < batchCount; i++) {
+      A_datas[i] = Ap->data;
+      B_datas[i] = Bp->data;
+      C_datas[i] = Cp->data;
+      A_offsets[i] = (Ap->offset + i * Ap->strides[0]) / elsize;
+      B_offsets[i] = (Bp->offset + i * Bp->strides[0]) / elsize;
+      C_offsets[i] = (Cp->offset + i * Cp->strides[0]) / elsize;
+    }
+
+    switch (C->typecode) {
+      case GA_HALF:
+        err = gpublas_hgemmBatch(o, transA, transB, m, n, k, (float)alpha,
```

Comment:
It might stop failing in the future.  Or it might work for OpenCL.

I would prefer to keep it as-is so that if someone in the future want to implement this they don't have to chase conditions up and down the stack.